include/deal.II/base/vectorization.h

```diff
@@ -805,7 +807,9 @@ class VectorizedArray<double>
     // is a bitwise operation so the data type does not matter)
     __m512d         mask = _mm512_set1_pd(-0.);
     VectorizedArray res;
-    res.data = (__m512d)_mm512_andnot_epi64((__m512i)mask, (__m512i)data);
+    res.data = reinterpret_cast<__m512d>(
```

Comment:
This doesn't look like a `reinterpret_cast` to me. It's not a pointer cast, but an object cast, no?

Reply:
I get 
```
/export/home/darndt/dealii/include/deal.II/base/vectorization.h:812:59: error: invalid static_cast from type ‘__m512i {aka __vector(8) long long int}’ to type ‘__m512d {aka __vector(8) double}’
```
with a `static_cast`.

Reply:
The above comment also says that we are essentially ignoring types here and just rely on the fact that the operations are performed bitwise.

Reply:
I did not know that that was possible at all. I thought `reinterpret_cast` was only for pointers. But https://en.cppreference.com/w/cpp/language/reinterpret_cast point 6 suggests that it also works for nonpointer types. (There is question about aliasing rules that we should probably ignore here.)

But this raises the question of what the original `(__m512d)expression` actually did. Isn't this the equivalent of a `static_cast`? How does C define this?

Reply:
I'll note that the referenced page has no examples where the type cast to is not a pointer. This makes me uneasy...

Reply:
Note that the compiler will not generate an assembly instruction from this line at all. It merely uses the register `zmm1` first for an integer operation (`andnot`) and then hands it back to the double array. I guess the compiler directly maps this to the memory type and thus treats it like a pointer - it really is a different interpretation of what the data in a segment of memory is.

Reply:
According to https://en.cppreference.com/w/cpp/language/explicit_cast the C-style cast previously used was equivalent to `reinterpret_cast` at least (since a `static_cast` didn't succeed and it is also not a candidate for a `const_cast`).

Reply:
We've settled the matter, but I wanted to look whether I can find anything more about `reinterpret_cast` being used for non-pointer types. Interestingly, there is essentially nothing one can find on the internet in this regard :-)